### PR TITLE
add iw-opd distillation

### PR DESCRIPTION
**Pull request by kashif**
## Summary

  Adds IW-OPD as an optional objective for the experimental distillation trainer.

  The new path uses sampled-token teacher and rollout logprobs to build the detached IW-OPD advantage, including cached vLLM rollout logprobs when  `use_vllm=True`.

  ## Tests

  - `make precommit`
  - `pytest tests/experimental/test_distillation_trainer.py -q`

Fixes # (issue)

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

## AI writing disclosure

We welcome the use of AI tools to help with contributions. For transparency and to help us improve our review process, please indicate the level of AI involvement in this PR.

- [ ] No AI usage: the PR was written entirely by a human.
- [x] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [ ] AI-generated: the PR was mostly or fully generated by an AI tool.

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag members/contributors who may be interested in your PR.

---

> [!NOTE]
> **Medium Risk**
> Changes on-policy loss routing and vLLM generation/buffer behavior; misconfiguration is mostly blocked by new validations, but IW-OPD training still depends on correct rollout logprobs and full teacher token coverage.
> 
> **Overview**
> Adds **Importance-Weighted On-Policy Distillation (IW-OPD)** as an optional objective on experimental `DistillationTrainer`, alongside the existing generalized JSD/KL path (`distillation_objective="jsd"`).
> 
> **Config:** New `distillation_objective` (`"jsd"` | `"iw_opd"`), `iw_opd_gamma`, and `iw_opd_epsilon`, with validation that IW-OPD requires `lmbda=1.0`, `reverse_kl_top_1_mode="sampled"`, no Liger kernel, and `vllm_sync_frequency=1` when `use_vllm=True`.
> 
> **Trainer:** `_compute_iw_opd_loss` builds detached OPD advantages from teacher vs rollout log-probs, applies prefix importance weights from accumulated teacher–student drift, and logs `iw_opd/*` metrics. `compute_loss` branches to IW-OPD for local and teacher-server paths (stricter teacher coverage on the server). vLLM rollouts request sampled-token logprobs, which are cached in the generation buffer as `rollout_logprobs` aligned with `labels`.
> 
> **Docs/tests:** User guide and paper index for IW-OPD; unit and integration tests for config guards, loss math, buffer layout, training, and server path.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit a69afce6abc2d7914e82b044d6547164b5450ca4. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/experimental/distillation/distillation_trainer.py [hunk 1]
```diff
@@ -1529,7 +1627,20 @@ def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=N
...
+                teacher_log_probs = F.log_softmax(teacher_logits / self.temperature, dim=-1)
+                teacher_actual_logprobs = teacher_log_probs.gather(
+                    dim=-1, index=completion_tokens.unsqueeze(-1)
+                ).squeeze(-1)
+                loss = self._compute_iw_opd_loss(
+                    student_logits=student_logits,
+                    completion_tokens=completion_tokens,
+                    labels=labels,
+                    teacher_actual_logprobs=teacher_actual_logprobs,
+                    rollout_logprobs=inputs.get("rollout_logprobs"),
```
Comment (cmpatino) [inline trl/experimental/distillation/distillation_trainer.py]: `rollout_logprobs` is completion-relative, but the student/teacher/label tensors are sliced from `_compute_prompt_length`. This introduces a misalignment between the rollouts and the labels.

I'm thinking we can store the rollouts it at full-sequence width in `_store_completions_in_buffer` and slice it with `[:, prompt_length:]` like we do for `labels`

Comment (kashif) [reply trl/experimental/distillation/distillation_trainer.py]: Good catch, fixed in 0e6cd0c0: rollout logprobs are now stored at full sequence width in _store_completions_in_buffer (zeros over the prompt) and sliced with [:, prompt_length:] alongside labels in compute_loss, so they stay aligned per row even with mixed prompt lengths. Added a regression check with different-length prompts to test_iw_opd_uses_cached_rollout_logprobs.

**Review by cmpatino — APPROVED** [review]
